Fix #37: Add unit tests for upload_file, and fix issues found therein

teodesian opened the pull request:
no_content_success was set incorrectly for the upload_file method.  Changing it to 0 resolved the lack of path being returned.

I also noted that selenium inconsistently returned the path to the file; sometimes the dirname, sometimes the basename.  Added tests for those cases, and implemented a workaround to always return the full path to the uploaded file.

gempesaw commented:
Heh, no worries about the mocks. Definitely a pain to set up, we can get them updated right before releasing to cpan.

gempesaw commented on t/01-driver.t:
```diff
@@ -450,6 +450,12 @@ STORAGE: {
+UPLOAD: {
+    #Webdriver only returns the full filename if there isn't any path components in it, so test both cases
+    like( $driver->upload_file('uploadTest'),qr/uploadTest$/,'upload_file returns FULL path to the file: cwd');
+    like( $driver->upload_file('t/uploadTest'),qr/uploadTest$/,'upload_file returns FULL path to the file: subdir');
```
Am I missing something about the return of `upload_file` that makes these tests sufficient? Should the expected have the full path in it somewhere?

teodesian replied:
These tests catch 2 particular problems:

1. The current behavior is to return 1 rather than the path at all.
2. If I pass 'somefile' as the argument, selenium returns /tmp/uuid/path/to/file/somefile.
    If I pass subdir/somefile as the argument, selenium returns /tmp/uuid/path/to/file/subdir, and omits the filename.  Rather than document this, I thought it was appropriate to make the upload_file method "do the right thing".

The test should not necesarily be specific beyond the filename; the path is in a temporary directory, which will vary across platforms; even File::Spec->tmpdir is not reliable in my experience.  Furthermore, it uses a randomly generated UUID which I cannot know beforehand in the path.

Probably the only more specificity the test could have would be checking that the session ID is a part of the path somewhere...

gempesaw replied:
Ah okay so it is addressing the weird case where it doesn't actually hand us back the filename. Thanks for the clarification.

gempesaw commented:
Oh, also, thanks much for the contribution :)

gempesaw commented on uploadTest:
```diff
@@ -0,0 +1 @@
+test
```
Oh, I didn't recognize this the first time around. Can we handle setup and cleanup for these uploadTest files in the normal running of `01-Driver.t` so they're not actually in the repository? 

In a passing test run it'd be great if the `UPLOAD: { }` section created both of these files, used them in the tests, and then deleted them afterwards. This is perhaps slightly complicated because we want a file in the current working dir for one of the test cases, and not just in a /tmp folder, hummmm.

teodesian replied:
Hmm, as long as it doesn't change the base64encode of the file; that might make a reliable recording for the mock impossible.

I'll give it a shot since it's a bit cleaner this way.

teodesian replied:
Regrettably, it appears that file ctime/mtime has an effect of the encoded output, making a reliable mock impossible with this approach.  That said, the ctime will be entirely different when run on a smoker for these files too;  This would explain why travis failed this PR.

The only other option for testing is to pass the zipped/base64'd string directly to upload_file.  I consider this an acceptable compromise, and will re-factor the upload_file method to optionally accept this sort of data.

teodesian commented:
I've modified upload_file such that it can pass a file's raw encode, and as such be mockable reliably.  I changed 01-driver.t to pass such. The test should pass and have reliable mocks able to be generated.

I thought that going any further for testing purposes (allowing passing of strings, false file attrs, etc) would be overkill, so here's hoping it's adequate.

The only concern at this point is what normally comes along with mocks and 'testing mode' -- the default use case (passing some filename) is now explicitly untested...but I can't think of a way that I could actually test that without some kind of equally grotesque evasion of the OS.

That said, I manually verified that code path was OK.  Perhaps I'll add some negative testing (check that it dies on nonexistant files, etc), to just make sure the path is at least gone down when we get expected input.

gempesaw commented:
Oh, I completely forgot about the annoying interactions between b64 encoding everything and the way our mocks are handled. Well, thanks for gracefully dealing with the extra complication. I'll merge this in momentarily.

gempesaw commented:
hm, I rebased your branch on to master and that seems to have duped the PR interface. Well, this is in master; thanks again!

gempesaw commented:
Thanks! Now on cpan as [v0.23](https://metacpan.org/release/GEMPESAW/Selenium-Remote-Driver-0.23)
